src/NUnitFramework/framework/Constraints/DictionaryContainsKeyConstraint.cs

```diff
@@ -65,32 +71,138 @@ public override string Description
         {
-            var dictionary = ConstraintUtils.RequireActual<IDictionary>(actual, nameof(actual));
+            if (_deprecatedFlag)
+            {
+                var dictionary = ConstraintUtils.RequireActual<IDictionary>(actual, nameof(actual));
+                foreach (object obj in dictionary.Keys)
+                    if (ItemsEqual(obj, Expected))
+                        return true;
+
+                return false;
+            }
 
-            foreach (object obj in dictionary.Keys)
-                if (ItemsEqual(obj, Expected))
-                    return true;
+            var method = GetContainsKeyMethod(actual);
+            if (method != null)
+                return (bool)method.Invoke(actual, new[] { Expected });
 
-            return false;
+            throw new ArgumentException("Not a collection supporting ContainsKey method.");
         }
 
         /// <summary>
         /// Flag the constraint to use the supplied predicate function
         /// </summary>
         /// <param name="comparison">The comparison function to use.</param>
         /// <returns>Self.</returns>
+        [Obsolete("Deprecated, use Does.ContainKey")]
         public DictionaryContainsKeyConstraint Using<TCollectionType, TMemberType>(Func<TCollectionType, TMemberType, bool> comparison)
         {
             // reverse the order of the arguments to match expectations of PredicateEqualityComparer
             Func<TMemberType, TCollectionType, bool> invertedComparison = (actual, expected) => comparison.Invoke(expected, actual);
 
+            _deprecatedFlag = true;
             base.Using(EqualityAdapter.For(invertedComparison));
             return this;
         }
+
+        /// <summary>
+        /// Flag the constraint to use the supplied Comparison object.
+        /// </summary>
+        /// <param name="comparer">The IComparer object to use.</param>
+        /// <returns>Self.</returns>
+        [Obsolete("Deprecated, use Does.ContainKey")]
+        public new CollectionItemsEqualConstraint Using<T>(Comparison<T> comparer)
+        {
+            _deprecatedFlag = true;
+            return base.Using(comparer);
+        }
+
+        /// <summary>
+        /// Checks if the key is contained in a "keyed item container".
+        /// </summary>
+        /// <param name="actual">Keyed container.</param>
+        /// <returns>method to call.</returns>
+        private MethodInfo GetContainsKeyMethod(object actual)
+        {
+            if (actual == null) throw new ArgumentNullException(nameof(actual));
+            var instanceType = actual.GetType();
+
+            var method = FindContainsKeyMethod(instanceType)
+                         ?? instanceType
+                            .GetInterfaces()
+                            .Concat(GetBaseTypes(instanceType))
+                            .Select(FindContainsKeyMethod)
+                            .FirstOrDefault(m => m != null);
+
```

Comment:
Thanks for bringing that to my attention. I'm really confused why [this method](https://github.com/nunit/nunit/blob/46b62b6dc4087efd0b4758ea039f3e07e7a7b911/src/NUnitFramework/framework/Internal/TypeHelper.cs#L316-L338) exists because Type.GetInterfaces returns all interfaces implemented by the type, which by definition includes all interfaces implemented by all base types. For example, try this in LINQPad:


<details>

```cs
void Main()
{
    typeof(B).GetInterfaces().Dump();
}

public class A : IEquatable<A>
{
    public bool Equals(A other)
    {
        throw new NotImplementedException();
    }
}

public class B : A, IComparable<B>
{
    public int CompareTo(B other)
    {
        throw new NotImplementedException();
    }
}
```

</details>

Reply:
I was confused at first, but `GetDeclaredInterfaces` only returns interfaces implemented by the most-derived type and intentionally excludes all interfaces implemented by base types. We actually want `Type.GetInterfaces()` here so that we catch all interface implementations.